File: TravelWise-backend/src/main/java/com/example/travelwise/entity/Offer.java
package com.example.travelwise.entity;

import jakarta.persistence.*;
import lombok.Data;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

@Data
@Entity
public class Offer {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String title;
    @Column(length = 2000)
    private String description;
    private String location;
    private Double price;
    private LocalDate startDate;
    private LocalDate endDate;
    private Integer availableSpots;
    @ManyToOne
    @JoinColumn(name="category_id")
    private Category category;
    @OneToMany(mappedBy = "offer", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Image> images = new ArrayList<>();

    @Convert(converter = StringListConverter.class)
    private List<String> tags = new ArrayList<>();
}

class StringListConverter implements AttributeConverter<List<String>, String> {

    @Override
    public String convertToDatabaseColumn(List<String> attribute) {
        if (attribute == null || attribute.isEmpty()) {
            return "{}";
        }
        return "{" + attribute.stream().map(tag -> "\"" + tag + "\"").collect(Collectors.joining(",")) + "}";
    }

    @Override
    public List<String> convertToEntityAttribute(String dbData) {
        if (dbData == null || dbData.isEmpty()) {
            return List.of();
        }
        return Arrays.asList(dbData.replaceAll("[{}\"]", "").split(","));
    }
}

File: TravelWise-frontend/src/app/models/offer.ts
export class Offer {
  id?: number;
  title: string;
  description: string;
  location: string;
  price: number;
  startDate: string;
  endDate: string;
  availableSpots: number;
  categoryId: number;
  imageUrls?: string[];

  constructor(title: string, description: string, location: string, price: number, startDate: string, endDate: string, availableSpots: number, categoryId: number) {
    this.title = title;
    this.description = description;
    this.location = location;
    this.price = price;
    this.startDate = startDate;
    this.endDate = endDate;
    this.availableSpots = availableSpots;
    this.categoryId = categoryId;
  }
}
